File: frontend/packages/stdb/src/wire-codec.ts
/**
 * Target-neutral compact operation codec.
 *
 * The contract uses lossless JSON values at the BFF boundary: unsigned
 * integers and timestamps are decimal strings, identities are canonical
 * lower-case hex, and SATS values are only introduced by this codec.  Keeping
 * this module descriptor-driven makes it suitable for generated operations and
 * resources without maintaining a reducer-specific serializer.
 */

export type CompactType = {
  kind: string
  [key: string]: unknown
}

export type CompactCodecCase = {
  name: string
  type: CompactType
  input?: unknown
  wire?: unknown
  error?: string
}

const MAX_U64 = 18_446_744_073_709_551_615n
const MAX_SAFE = BigInt(Number.MAX_SAFE_INTEGER)
const MIN_I64 = -(1n << 63n)
const MAX_I64 = (1n << 63n) - 1n

function fail(code: string, message: string): never {
  throw new Error(`compact-codec:${code}: ${message}`)
}

function objectValue(value: unknown, code: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    fail(code, "expected an object")
  }
  return value as Record<string, unknown>
}

function decimal(value: unknown, code: string, min: bigint, max: bigint): string {
  let parsed: bigint
  if (typeof value === "bigint") {
    parsed = value
  } else if (typeof value === "string" && (value === "0" || /^-?[1-9]\d*$/.test(value))) {
    try {
      parsed = BigInt(value)
    } catch {
      fail(code, "invalid decimal")
    }
  } else if (typeof value === "number" && Number.isSafeInteger(value)) {
    parsed = BigInt(value)
  } else {
    fail(code, "expected a decimal string or safe integer")
  }
  if (parsed < min || parsed > max) fail(code, "integer is out of range")
  return parsed.toString()
}

function identity(value: unknown): string {
  const obj = value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : undefined
  if (obj && (Object.keys(obj).length !== 1 || !("__identity__" in obj))) {
    fail("identity", "malformed identity object")
  }
  const raw = obj && typeof obj["__identity__"] === "string"
    ? obj["__identity__"]
    : value
  if (typeof raw !== "string") fail("identity", "expected a 64-character hex string")
  const hex = raw.replace(/^0x/i, "")
  if (!/^[0-9a-fA-F]{64}$/.test(hex)) fail("identity", "expected a 64-character hex string")
  return `0x${hex.toLowerCase()}`
}

function timestamp(value: unknown): string {
  const obj = objectValue(value, "timestamp")
  if (Object.keys(obj).length !== 1 || !("microsSinceUnixEpoch" in obj)) {
    fail("timestamp", "malformed timestamp object")
  }
  return decimal(obj["microsSinceUnixEpoch"], "timestamp", MIN_I64, MAX_I64)
}

function variantKey(tag: string): string {
  return tag.length === 0 ? "" : tag.charAt(0).toLowerCase() + tag.slice(1)
}

function variants(type: CompactType): Record<string, unknown> {
  const value = type["variants"]
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    fail("descriptor", "enum variants must be an object")
  }
  return value as Record<string, unknown>
}

function fields(type: CompactType): Record<string, Record<string, unknown>> {
  const raw = type["fields"]
  if (!Array.isArray(raw)) fail("descriptor", "struct fields must be an array")
  const result: Record<string, Record<string, unknown>> = {}
  for (const field of raw) {
    const item = objectValue(field, "descriptor")
    if (typeof item["name"] !== "string" || typeof item["wire"] !== "string") {
      fail("descriptor", "struct fields require name and wire")
    }
    if (result[item["name"] as string]) fail("descriptor", "duplicate struct field")
    result[item["name"] as string] = item
  }
  return result
}

function aliases(field: Record<string, unknown>): string[] {
  const value = field["aliases"]
  if (value === undefined) return []
  if (!Array.isArray(value) || value.some((item) => typeof item !== "string")) {
    fail("descriptor", "field aliases must be strings")
  }
  return value as string[]
}

/** Encode canonical operation/resource values to strict SATS JSON wire values. */
export function encodeCompact(type: CompactType, value: unknown): unknown {
  switch (type.kind) {
    case "alias":
      return encodeCompact(objectValue(type["target"], "descriptor") as CompactType, value)
    case "u64":
      return decimal(value, "u64", 0n, MAX_U64)
    case "timestamp":
      return { __timestamp_micros_since_unix_epoch__: timestamp(value) }
    case "identity":
      return { __identity__: identity(value) }
    case "string":
      if (typeof value !== "string") fail("string", "expected a string")
      return value
    case "bool":
      if (typeof value !== "boolean") fail("bool", "expected a boolean")
      return value
    case "option": {
      const inner = objectValue(type["inner"], "descriptor") as CompactType
      return value === null || value === undefined
        ? { none: [] }
        : { some: encodeCompact(inner, value) }
    }
    case "array": {
      if (!Array.isArray(value)) fail("array", "expected an array")
      const inner = objectValue(type["items"], "descriptor") as CompactType
      return value.map((item) => encodeCompact(inner, item))
    }
    case "enum": {
      const input = objectValue(value, "enum")
      if (typeof input["tag"] !== "string") fail("enum", "tag is required")
      const tag = input["tag"]
      const definition = variants(type)[tag]
      if (definition === undefined) fail("enum", `unknown tag ${tag}`)
      const keys = Object.keys(input)
      const hasValue = Object.prototype.hasOwnProperty.call(input, "value")
      if (definition === null) {
        if (hasValue || keys.length !== 1) fail("enum", "unit variant cannot carry a value")
        return { [variantKey(tag)]: [] }
      }
      if (!hasValue || keys.length !== 2) fail("enum", "payload variant requires value")
      return { [variantKey(tag)]: encodeCompact(objectValue(definition, "descriptor") as CompactType, input["value"]) }
    }
    case "struct": {
      const input = objectValue(value, "struct")
      const known = fields(type)
      const output: Record<string, unknown> = {}
      for (const [name, field] of Object.entries(known)) {
        const accepted = [name, ...aliases(field)]
        const present = accepted.filter((key) => Object.prototype.hasOwnProperty.call(input, key))
        if (present.length > 1) fail("alias", `multiple aliases supplied for ${name}`)
        if (present.length === 1) {
          output[field["wire"] as string] = encodeCompact(
            objectValue(field["type"], "descriptor") as CompactType,
            input[present[0]!],
          )
        }
      }
      const accepted = new Set(Object.entries(known).flatMap(([name, field]) => [name, ...aliases(field)]))
      for (const key of Object.keys(input)) {
        if (!accepted.has(key)) fail("field", `unknown field ${key}`)
      }
      return output
    }
    default:
      fail("descriptor", `unsupported type ${type.kind}`)
  }
}

/** Decode strict SATS JSON wire values to canonical, lossless contract values. */
export function decodeCompact(type: CompactType, value: unknown): unknown {
  switch (type.kind) {
    case "alias":
      return decodeCompact(objectValue(type["target"], "descriptor") as CompactType, value)
    case "u64":
      return decimal(value, "u64", 0n, MAX_U64)
    case "timestamp": {
      const input = objectValue(value, "timestamp")
      if (Object.keys(input).length !== 1 || !("__timestamp_micros_since_unix_epoch__" in input)) {
        fail("timestamp", "malformed wire timestamp")
      }
      return { microsSinceUnixEpoch: decimal(input["__timestamp_micros_since_unix_epoch__"], "timestamp", MIN_I64, MAX_I64) }
    }
    case "identity": {
      const input = objectValue(value, "identity")
      if (Object.keys(input).length !== 1 || typeof input["__identity__"] !== "string") fail("identity", "malformed wire identity")
      return identity(input["__identity__"])
    }
    case "string":
      if (typeof value !== "string") fail("string", "expected a string")
      return value
    case "bool":
      if (typeof value !== "boolean") fail("bool", "expected a boolean")
      return value
    case "option": {
      const input = objectValue(value, "option")
      if (Object.keys(input).length !== 1) fail("option", "option must have one variant")
      if ("none" in input) {
        if (!Array.isArray(input["none"]) || (input["none"] as unknown[]).length !== 0) fail("option", "malformed none payload")
        return null
      }
      if (!("some" in input)) fail("option", "unknown option variant")
      return decodeCompact(objectValue(type["inner"], "descriptor") as CompactType, input["some"])
    }
    case "array": {
      if (!Array.isArray(value)) fail("array", "expected an array")
      return value.map((item) => decodeCompact(objectValue(type["items"], "descriptor") as CompactType, item))
    }
    case "enum": {
      const input = objectValue(value, "enum")
      if (Object.keys(input).length !== 1) fail("enum", "enum must have one variant")
      const wire = Object.keys(input)[0]!
      const tag = Object.keys(variants(type)).find((candidate) => variantKey(candidate) === wire)
      if (!tag) fail("enum", `unknown wire tag ${wire}`)
      const definition = variants(type)[tag]
      const payload = input[wire]
      if (definition === null) {
        if (!Array.isArray(payload) || payload.length !== 0) fail("enum", "malformed unit payload")
        return { tag }
      }
      return { tag, value: decodeCompact(objectValue(definition, "descriptor") as CompactType, payload) }
    }
    case "struct": {
      const input = objectValue(value, "struct")
      const output: Record<string, unknown> = {}
      for (const [name, field] of Object.entries(fields(type))) {
        const wire = field["wire"] as string
        if (Object.prototype.hasOwnProperty.call(input, wire)) {
          output[name] = decodeCompact(objectValue(field["type"], "descriptor") as CompactType, input[wire])
        }
      }
      const wires = new Set(Object.values(fields(type)).map((field) => field["wire"] as string))
      for (const key of Object.keys(input)) if (!wires.has(key)) fail("field", `unknown wire field ${key}`)
      return output
    }
    default:
      fail("descriptor", `unsupported type ${type.kind}`)
  }
}

// Prevent accidental removal of the range constants when the implementation
// is tree-shaken in a generated-only build; these are part of the contract.
export const COMPACT_CODEC_VERSION = 1
export const COMPACT_CODEC_MAX_SAFE_INTEGER = MAX_SAFE

File: crates/stdb-client/src/wire_codec.rs
//! Strict descriptor-driven codec for compact operation/resource contracts.
//!
//! Contract values use lossless JSON (`u64` and timestamp values are decimal
//! strings) and are translated to the SATS JSON shape only at the transport
//! boundary.  The descriptor is intentionally JSON-shaped so the generated
//! contracts crate can provide the same table to TypeScript and Rust.

use serde_json::{Map, Value};
use std::fmt;

const MAX_U64: u128 = 18_446_744_073_709_551_615;
const MIN_I64: i128 = -9_223_372_036_854_775_808;
const MAX_I64: i128 = 9_223_372_036_854_775_807;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CompactCodecError(String);

impl CompactCodecError {
    fn new(code: &str, message: impl Into<String>) -> Self {
        Self(format!("compact-codec:{code}: {}", message.into()))
    }

    /// Stable machine-readable class (`u64`, `identity`, `field`, ...).
    pub fn code(&self) -> &str {
        self.0
            .strip_prefix("compact-codec:")
            .and_then(|value| value.split_once(':'))
            .map(|(code, _)| code)
            .unwrap_or("unknown")
    }
}

impl fmt::Display for CompactCodecError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl std::error::Error for CompactCodecError {}

type Result<T> = std::result::Result<T, CompactCodecError>;

fn object<'a>(value: &'a Value, code: &str) -> Result<&'a Map<String, Value>> {
    value
        .as_object()
        .ok_or_else(|| CompactCodecError::new(code, "expected an object"))
}

fn type_kind<'a>(ty: &'a Value) -> Result<&'a str> {
    object(ty, "descriptor")?
        .get("kind")
        .and_then(Value::as_str)
        .ok_or_else(|| CompactCodecError::new("descriptor", "type kind is required"))
}

fn decimal(value: &Value, code: &str, min: i128, max: i128) -> Result<String> {
    let raw = match value {
        Value::String(value) => value.clone(),
        Value::Number(value) => value.to_string(),
        _ => {
            return Err(CompactCodecError::new(
                code,
                "expected a decimal string or integer",
            ))
        }
    };
    let digits = raw.strip_prefix('-').unwrap_or(&raw);
    if raw.is_empty()
        || digits.is_empty()
        || (raw.starts_with('-') && digits == "0")
        || (digits.len() > 1 && digits.starts_with('0'))
        || !digits.chars().all(|c| c.is_ascii_digit())
    {
        return Err(CompactCodecError::new(code, "invalid decimal"));
    }
    let parsed = raw
        .parse::<i128>()
        .map_err(|_| CompactCodecError::new(code, "invalid decimal"))?;
    if parsed < min || parsed > max {
        return Err(CompactCodecError::new(code, "integer is out of range"));
    }
    Ok(parsed.to_string())
}

fn identity(value: &Value) -> Result<String> {
    let raw = if let Some(object) = value.as_object() {
        if object.len() != 1 || !object.contains_key("__identity__") {
            return Err(CompactCodecError::new(
                "identity",
                "malformed identity object",
            ));
        }
        object
            .get("__identity__")
            .and_then(Value::as_str)
            .ok_or_else(|| CompactCodecError::new("identity", "expected a hex string"))?
    } else {
        value
            .as_str()
            .ok_or_else(|| CompactCodecError::new("identity", "expected a hex string"))?
    };
    let hex = raw
        .strip_prefix("0x")
        .or_else(|| raw.strip_prefix("0X"))
        .unwrap_or(raw);
    if hex.len() != 64 || !hex.bytes().all(|byte| byte.is_ascii_hexdigit()) {
        return Err(CompactCodecError::new(
            "identity",
            "expected a 64-character hex string",
        ));
    }
    Ok(format!("0x{}", hex.to_ascii_lowercase()))
}

fn variant_key(tag: &str) -> String {
    let mut chars = tag.chars();
    chars
        .next()
        .map(|first| first.to_ascii_lowercase().to_string() + chars.as_str())
        .unwrap_or_default()
}

fn variants<'a>(ty: &'a Value) -> Result<&'a Map<String, Value>> {
    object(ty, "descriptor")?
        .get("variants")
        .and_then(Value::as_object)
        .ok_or_else(|| CompactCodecError::new("descriptor", "enum variants must be an object"))
}

fn fields<'a>(ty: &'a Value) -> Result<&'a Vec<Value>> {
    object(ty, "descriptor")?
        .get("fields")
        .and_then(Value::as_array)
        .ok_or_else(|| CompactCodecError::new("descriptor", "struct fields must be an array"))
}

fn field<'a>(value: &'a Value, key: &str) -> Result<&'a Map<String, Value>> {
    let field = object(value, "descriptor")?;
    if field.get("name").and_then(Value::as_str).is_none()
        || field.get("wire").and_then(Value::as_str).is_none()
        || field.get("type").is_none()
    {
        return Err(CompactCodecError::new(
            "descriptor",
            format!("struct field {key} requires name, wire, and type"),
        ));
    }
    Ok(field)
}

fn aliases(field: &Map<String, Value>) -> Result<Vec<&str>> {
    let Some(value) = field.get("aliases") else {
        return Ok(Vec::new());
    };
    value
        .as_array()
        .ok_or_else(|| CompactCodecError::new("descriptor", "field aliases must be an array"))?
        .iter()
        .map(|value| {
            value.as_str().ok_or_else(|| {
                CompactCodecError::new("descriptor", "field aliases must be strings")
            })
        })
        .collect()
}

fn nested_type<'a>(map: &'a Map<String, Value>, key: &str) -> Result<&'a Value> {
    map.get(key)
        .ok_or_else(|| CompactCodecError::new("descriptor", format!("missing {key} type")))
}

/// Encode a canonical contract value as strict SATS JSON.
pub fn encode_compact(ty: &Value, value: &Value) -> Result<Value> {
    match type_kind(ty)? {
        "alias" => encode_compact(nested_type(object(ty, "descriptor")?, "target")?, value),
        "u64" => Ok(Value::String(decimal(value, "u64", 0, MAX_U64 as i128)?)),
        "timestamp" => {
            let input = object(value, "timestamp")?;
            if input.len() != 1 || !input.contains_key("microsSinceUnixEpoch") {
                return Err(CompactCodecError::new(
                    "timestamp",
                    "malformed timestamp object",
                ));
            }
            Ok(serde_json::json!({
                "__timestamp_micros_since_unix_epoch__": decimal(
                    &input["microsSinceUnixEpoch"],
                    "timestamp",
                    MIN_I64,
                    MAX_I64,
                )?
            }))
        }
        "identity" => Ok(serde_json::json!({ "__identity__": identity(value)? })),
        "string" => value
            .as_str()
            .map(|value| Value::String(value.to_owned()))
            .ok_or_else(|| CompactCodecError::new("string", "expected a string")),
        "bool" => value
            .as_bool()
            .map(Value::Bool)
            .ok_or_else(|| CompactCodecError::new("bool", "expected a boolean")),
        "option" => {
            let inner = nested_type(object(ty, "descriptor")?, "inner")?;
            if value.is_null() {
                Ok(serde_json::json!({ "none": [] }))
            } else {
                Ok(serde_json::json!({ "some": encode_compact(inner, value)? }))
            }
        }
        "array" => value
            .as_array()
            .ok_or_else(|| CompactCodecError::new("array", "expected an array"))?
            .iter()
            .map(|item| encode_compact(nested_type(object(ty, "descriptor")?, "items")?, item))
            .collect::<Result<Vec<_>>>()
            .map(Value::Array),
        "enum" => {
            let input = object(value, "enum")?;
            let tag = input
                .get("tag")
                .and_then(Value::as_str)
                .ok_or_else(|| CompactCodecError::new("enum", "tag is required"))?;
            let definition = variants(ty)?
                .get(tag)
                .ok_or_else(|| CompactCodecError::new("enum", format!("unknown tag {tag}")))?;
            let mut result = Map::new();
            if definition.is_null() {
                if input.len() != 1 {
                    return Err(CompactCodecError::new(
                        "enum",
                        "unit variant cannot carry a value",
                    ));
                }
                result.insert(variant_key(tag), Value::Array(Vec::new()));
            } else {
                let payload = input.get("value").ok_or_else(|| {
                    CompactCodecError::new("enum", "payload variant requires value")
                })?;
                if input.len() != 2 {
                    return Err(CompactCodecError::new("enum", "unknown enum field"));
                }
                result.insert(variant_key(tag), encode_compact(definition, payload)?);
            }
            Ok(Value::Object(result))
        }
        "struct" => {
            let input = object(value, "struct")?;
            let mut result = Map::new();
            let mut accepted = Vec::new();
            for raw_field in fields(ty)? {
                let field = field(raw_field, "")?;
                let name = field["name"].as_str().expect("validated field name");
                let wire = field["wire"].as_str().expect("validated field wire");
                accepted.push(name);
                let field_aliases = aliases(field)?;
                accepted.extend(field_aliases.iter().copied());
                let present: Vec<&str> = std::iter::once(name)
                    .chain(field_aliases.iter().copied())
                    .filter(|key| input.contains_key(*key))
                    .collect();
                if present.len() > 1 {
                    return Err(CompactCodecError::new(
                        "alias",
                        format!("multiple aliases supplied for {name}"),
                    ));
                }
                if let Some(input_key) = present.first() {
                    result.insert(
                        wire.to_owned(),
                        encode_compact(nested_type(field, "type")?, &input[*input_key])?,
                    );
                }
            }
            if input.keys().any(|key| !accepted.contains(&key.as_str())) {
                return Err(CompactCodecError::new("field", "unknown field"));
            }
            Ok(Value::Object(result))
        }
        other => Err(CompactCodecError::new(
            "descriptor",
            format!("unsupported type {other}"),
        )),
    }
}

/// Decode strict SATS JSON as a canonical, lossless contract value.
pub fn decode_compact(ty: &Value, value: &Value) -> Result<Value> {
    match type_kind(ty)? {
        "alias" => decode_compact(nested_type(object(ty, "descriptor")?, "target")?, value),
        "u64" => Ok(Value::String(decimal(value, "u64", 0, MAX_U64 as i128)?)),
        "timestamp" => {
            let input = object(value, "timestamp")?;
            if input.len() != 1 || !input.contains_key("__timestamp_micros_since_unix_epoch__") {
                return Err(CompactCodecError::new(
                    "timestamp",
                    "malformed wire timestamp",
                ));
            }
            Ok(
                serde_json::json!({ "microsSinceUnixEpoch": decimal(&input["__timestamp_micros_since_unix_epoch__"], "timestamp", MIN_I64, MAX_I64)? }),
            )
        }
        "identity" => {
            let input = object(value, "identity")?;
            if input.len() != 1 || !input.contains_key("__identity__") {
                return Err(CompactCodecError::new(
                    "identity",
                    "malformed wire identity",
                ));
            }
            Ok(Value::String(identity(value)?))
        }
        "string" => value
            .as_str()
            .map(|value| Value::String(value.to_owned()))
            .ok_or_else(|| CompactCodecError::new("string", "expected a string")),
        "bool" => value
            .as_bool()
            .map(Value::Bool)
            .ok_or_else(|| CompactCodecError::new("bool", "expected a boolean")),
        "option" => {
            let input = object(value, "option")?;
            if input.len() != 1 {
                return Err(CompactCodecError::new(
                    "option",
                    "option must have one variant",
                ));
            }
            if let Some(none) = input.get("none") {
                if none.as_array().map_or(true, |items| !items.is_empty()) {
                    return Err(CompactCodecError::new("option", "malformed none payload"));
                }
                return Ok(Value::Null);
            }
            let some = input
                .get("some")
                .ok_or_else(|| CompactCodecError::new("option", "unknown option variant"))?;
            decode_compact(nested_type(object(ty, "descriptor")?, "inner")?, some)
        }
        "array" => value
            .as_array()
            .ok_or_else(|| CompactCodecError::new("array", "expected an array"))?
            .iter()
            .map(|item| decode_compact(nested_type(object(ty, "descriptor")?, "items")?, item))
            .collect::<Result<Vec<_>>>()
            .map(Value::Array),
        "enum" => {
            let input = object(value, "enum")?;
            if input.len() != 1 {
                return Err(CompactCodecError::new("enum", "enum must have one variant"));
            }
            let wire = input.keys().next().expect("one enum key");
            let (tag, definition) = variants(ty)?
                .iter()
                .find(|(tag, _)| variant_key(tag) == *wire)
                .ok_or_else(|| {
                    CompactCodecError::new("enum", format!("unknown wire tag {wire}"))
                })?;
            if definition.is_null() {
                if !input[wire].as_array().is_some_and(Vec::is_empty) {
                    return Err(CompactCodecError::new("enum", "malformed unit payload"));
                }
                return Ok(serde_json::json!({ "tag": tag }));
            }
            Ok(
                serde_json::json!({ "tag": tag, "value": decode_compact(definition, &input[wire])? }),
            )
        }
        "struct" => {
            let input = object(value, "struct")?;
            let mut result = Map::new();
            let mut wires = Vec::new();
            for raw_field in fields(ty)? {
                let field = field(raw_field, "")?;
                let name = field["name"].as_str().expect("validated field name");
                let wire = field["wire"].as_str().expect("validated field wire");
                wires.push(wire);
                if let Some(value) = input.get(wire) {
                    result.insert(
                        name.to_owned(),
                        decode_compact(nested_type(field, "type")?, value)?,
                    );
                }
            }
            if input.keys().any(|key| !wires.contains(&key.as_str())) {
                return Err(CompactCodecError::new("field", "unknown wire field"));
            }
            Ok(Value::Object(result))
        }
        other => Err(CompactCodecError::new(
            "descriptor",
            format!("unsupported type {other}"),
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::{decode_compact, encode_compact};
    use serde_json::Value;

    fn fixtures() -> Value {
        serde_json::from_str(include_str!(
            "../../../frontend/packages/stdb/src/wire-codec-fixtures.json"
        ))
        .expect("wire codec fixture JSON")
    }

    #[test]
    fn shared_fixture_corpus_encodes_and_decodes() {
        for fixture in fixtures()["cases"].as_array().expect("cases") {
            if fixture.get("error").is_some() {
                if fixture.get("wire").is_none() {
                    assert!(
                        encode_compact(&fixture["type"], &fixture["input"]).is_err(),
                        "{}",
                        fixture["name"]
                    );
                }
                if fixture.get("wire").is_some() {
                    assert!(
                        decode_compact(&fixture["type"], &fixture["wire"]).is_err(),
                        "{} wire",
                        fixture["name"]
                    );
                }
            } else {
                let wire = encode_compact(&fixture["type"], &fixture["input"]).expect("encode");
                assert_eq!(wire, fixture["wire"], "{}", fixture["name"]);
                let expected = fixture
                    .get("canonical")
                    .unwrap_or(&fixture["input"])
                    .clone();
                assert_eq!(
                    decode_compact(&fixture["type"], &wire).expect("decode"),
                    expected,
                    "{}",
                    fixture["name"]
                );
            }
        }
    }
}
